ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java

```diff
@@ -208,30 +208,48 @@ public void resourceChanged(IResourceChangeEvent event) {
 		IResourceDelta delta = event.getDelta();
 
 		try {
+			List<IFile> added = new ArrayList<>();
+			List<IFile> removed = new ArrayList<>();
 			delta.accept(delta1 -> {
-				int kind = delta1.getKind();
 				IResource resource = delta1.getResource();
-				if (resource instanceof IFile) {
+				if (resource.getType() == IResource.FILE) {
+					int kind = delta1.getKind();
 					if (kind == IResourceDelta.ADDED)
-						asyncStructureChanged((IFile) resource, true);
+						added.add((IFile) resource);
 					else if (kind == IResourceDelta.REMOVED)
-						asyncStructureChanged((IFile) resource, false);
+						removed.add((IFile) resource);
 					return false;
 				}
 				return true;
 			});
+			if (!added.isEmpty() || !removed.isEmpty()) {
+				asyncStructureChanged(added, removed);
+			}
 		} catch (CoreException e) {
 			PDEPlugin.logException(e);
 		}
 	}
 
-	private void asyncStructureChanged(final IFile file, final boolean added) {
-		if (editor == null || editor.getEditorSite() == null)
+	private void asyncStructureChanged(List<IFile> added, List<IFile> removed) {
+		if (monitoredFiles == null)
 			return;
-		Shell shell = editor.getEditorSite().getShell();
-		Display display = shell != null ? shell.getDisplay() : Display.getDefault();
-
-		display.asyncExec(() -> structureChanged(file, added));
+		try {
+			IProgressService progressService = PlatformUI.getWorkbench().getProgressService();
+			progressService.runInUI(progressService, monitor -> {
+				for (IFile file : added) {
```

Comment:
This looks better now (haven't tested), but **please** as the first task in the UI thread, check if the editor is still there.